File: 1043.partition-array-for-maximum-sum.go
// @leet start
func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func dfs(arr []int, k int, i int, memo []int) int {
	if memo[i] > 0 {
		return memo[i]
	}
	max := 0
	res := 0
	for j := i; j < i+k && j < len(arr); j++ {
		max = maxInt(max, arr[j])
		res = maxInt(res, max*(j-i+1)+dfs(arr, k, j+1, memo))
	}
	memo[i] = res
	return res
}

func maxSumAfterPartitioning(arr []int, k int) int {
	return dfs(arr, k, 0, make([]int, len(arr)+1))
}

// @leet end

File: 1043.partition-array-for-maximum-sum.kt
// @leet start
class Solution {
    fun maxSumAfterPartitioning(arr: IntArray, k: Int): Int {
      val cache = IntArray(arr.size + 1)
      fun dfs(i: Int): Int {
        if (cache[i] != 0) return cache[i]
        var max = 0
        var res = 0
        for (j in i until minOf(i + k, arr.size)) {
          max = maxOf(max, arr[j])
          res = maxOf(res, dfs(j + 1) + max * (j - i + 1))
        }
        cache[i] = res
        return res
      }
      return dfs(0)
    }
}
// @leet end
